### PR TITLE
fix tag byte overlap issue #520

### PR DESCRIPTION
**Pull request by DreamWuGit**
this PR try to fix issue #520

### CODE REVIEW

**Comment by lispc** [comment]
you may need to update the comments here too? https://github.com/appliedzkp/zkevm-circuits/blob/1b1456609cd6ac9eb54575cdb60a098af3bbe0e5/zkevm-circuits/src/state_circuit/lexicographic_ordering.rs#L16

**Comment by lispc** [comment]
@DreamWuGit if you change field tag to 5 bits, the total bits will not be 480 bits? 481 bits? Or reduce 1 bit from some other key?

**Comment by z2trillion** [comment]
For the proofs to verify and the tests to pass, you need to update `packed_tags` in line 282.

As written, the extra bit comes from the second most significant byte of the id. The maximum value of id we can handle after this change will only be 2^16 - 1.

**Comment by DreamWuGit** [comment]
> @DreamWuGit if you change field tag to 5 bits, the total bits will not be 480 bits? 481 bits? Or reduce 1 bit from some other key?

481 bits seem ok to me , will check with Mason if  he has another idea !

**Comment by lispc** [comment]
481 bits need 31 16bit limbs. And much refactor needed... Better keep 480bits i think

**Comment by z2trillion** [comment]
this still keeps it to 480 bits, so 2 field elements.

**Comment by DreamWuGit** [comment]
@miha-stopar  could you have a look ?

File: zkevm-circuits/src/state_circuit/lexicographic_ordering.rs [hunk 1]
```diff
@@ -302,25 +302,27 @@ impl<F: Field> Queries<F> {
...
+        limbs[0] = limbs[0].clone() + self.packed_tags() * (1u64 << 7).expr();
         limbs
     }
 }
 
 fn rw_to_be_limbs(row: &Rw) -> Vec<u16> {
+    let mut id = row.id().unwrap_or_default() as u32;
+    assert_eq!(id.to_be_bytes().len(), 4);
+    // check that the most significant 9 bits of id are 0, and pack tag and then
+    // field_tag into them.
```
Comment (miha-stopar) [inline zkevm-circuits/src/state_circuit/lexicographic_ordering.rs:316]: Perhaps add some short explanation why the most significant 9 bits are 0?

File: zkevm-circuits/src/state_circuit/lexicographic_ordering.rs [hunk 2]
```diff
@@ -302,25 +302,27 @@ impl<F: Field> Queries<F> {
             .chain(self.rw_counter_limbs.iter().rev())
             .cloned()
             .collect();
-        // most significant byte of id should be 0, so safe to overwrite it with packed
-        // tags.
-        limbs[0] = limbs[0].clone() + self.packed_tags() * (1u64 << 8).expr();
+        // most significant 9 bits of id are assumed be 0, so we can overwrite them with
+        // packed tags.
+        limbs[0] = limbs[0].clone() + self.packed_tags() * (1u64 << 7).expr();
```
Comment (miha-stopar) [inline zkevm-circuits/src/state_circuit/lexicographic_ordering.rs:307]: Perhaps add a hint 9 + 7 = 16 to be clear we are moving the value to the most significant 9 bits of the 16 bits?

File: zkevm-circuits/src/state_circuit/lexicographic_ordering.rs [hunk 3]
```diff
@@ -60,8 +60,8 @@ use std::ops::Mul;
 
 // Packing the field into 480 bits:
 //   4 bits for tag,
-// + 4 bits for field_tag // TODO: this actually needs 5 bits. Either reduce id
-// + 24 bits for id       // to 23 bits, or add diff_3 etc.
+// + 5 bits for field_tag
+// + 24 bits for id
```
Comment (miha-stopar) [inline zkevm-circuits/src/state_circuit/lexicographic_ordering.rs:64]: 23, right?

**Review by miha-stopar — APPROVED** [review]
Looks cool to me, I would just perhaps add a couple of comments to make it more easy to understand the idea.

File: zkevm-circuits/src/state_circuit/lexicographic_ordering.rs (hunk 3 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/state_circuit/lexicographic_ordering.rs:64]: right,  will update !! thanks

Comment (DreamWuGit) [reply zkevm-circuits/src/state_circuit/lexicographic_ordering.rs]: updated in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/523/commits/f8314b0ebe9246744344c56f02c89ab327ac83d3 thanks !

File: zkevm-circuits/src/state_circuit/lexicographic_ordering.rs (hunk 2 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/state_circuit/lexicographic_ordering.rs:307]: added comments in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/523/commits/f8314b0ebe9246744344c56f02c89ab327ac83d3 thanks !

File: zkevm-circuits/src/state_circuit/lexicographic_ordering.rs (hunk 1 above)
Comment (DreamWuGit) [reply zkevm-circuits/src/state_circuit/lexicographic_ordering.rs]: updated in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/523/commits/f8314b0ebe9246744344c56f02c89ab327ac83d3 thanks !

File: zkevm-circuits/src/state_circuit/lexicographic_ordering.rs [hunk 4]
```diff
@@ -316,25 +316,27 @@ impl<F: Field> Queries<F> {
...
+        limbs[0] = limbs[0].clone() + self.packed_tags() * (1u64 << 7).expr();
         limbs
     }
 }
 
 fn rw_to_be_limbs(row: &Rw) -> Vec<u16> {
+    let mut id = row.id().unwrap_or_default() as u32;
+    assert_eq!(id.to_be_bytes().len(), 4);
+    // id is u32, but it is at most 2^23 - 1, so the 9 most significant bits will be
+    // 0. We overwrite these 9 bits with the tag and field tag.
```
Comment (icemelon) [inline zkevm-circuits/src/state_circuit/lexicographic_ordering.rs:330]:
```suggestion
    // The max value of `id` is 2^23 - 1, so the 9 most significant bits should be
    // 0. We use these 9 bits to hold value of `tag` and `field_tag`.
```

**Review by icemelon — APPROVED** [review]
LGTM. minor revise on the comment.

**Pull request merged**